precompiles/erc20/approve.go

```diff
@@ -211,36 +222,82 @@ func (p Precompile) createAuthorization(ctx sdk.Context, grantee, granter common
 	return p.AuthzKeeper.SaveGrant(ctx, grantee.Bytes(), granter.Bytes(), authorization, &expiration)
 }
 
-func (p Precompile) updateAuthorization(ctx sdk.Context, grantee, granter common.Address, amount *big.Int, authorization *banktypes.SendAuthorization) error {
-	authorization.SpendLimit = sdk.Coins{{Denom: p.tokenPair.Denom, Amount: sdk.NewIntFromBigInt(amount)}}
+func (p Precompile) updateAuthorization(ctx sdk.Context, grantee, granter common.Address, amount *big.Int, authorization *banktypes.SendAuthorization, expiration *time.Time) error {
+	authorization.SpendLimit = updateOrAddCoin(authorization.SpendLimit, sdk.Coin{Denom: p.tokenPair.Denom, Amount: sdk.NewIntFromBigInt(amount)})
 	if err := authorization.ValidateBasic(); err != nil {
 		return err
 	}
 
-	expiration := ctx.BlockTime().Add(p.ApprovalExpiration)
+	return p.AuthzKeeper.SaveGrant(ctx, grantee.Bytes(), granter.Bytes(), authorization, expiration)
+}
 
-	return p.AuthzKeeper.SaveGrant(ctx, grantee.Bytes(), granter.Bytes(), authorization, &expiration)
+// updateOrAddCoin replaces the coin of the given denomination in the coins slice or adds it if it
+// does not exist yet.
+//
+// CONTRACT: Requires the coins struct to contain at most one coin of the given
+// denom.
+func updateOrAddCoin(coins sdk.Coins, coin sdk.Coin) sdk.Coins {
+	for idx, c := range coins {
+		if c.Denom == coin.Denom {
+			coins[idx] = coin
+			return coins
+		}
+	}
+
+	// NOTE: if no coin with the correct denomination is in the coins slice, we
+	// add it here.
+	return coins.Add(coin)
 }
 
-func (p Precompile) deleteAuthorization(ctx sdk.Context, grantee, granter common.Address) error {
-	return p.AuthzKeeper.DeleteGrant(ctx, grantee.Bytes(), granter.Bytes(), SendMsgURL)
+// removeSpendLimitOrDeleteAuthorization removes the spend limit for the given
+// token and updates the grant or deletes the authorization if no spend limit in another
+// denomination is set.
+func (p Precompile) removeSpendLimitOrDeleteAuthorization(ctx sdk.Context, grantee, granter common.Address, authorization authz.Authorization, expiration *time.Time) error {
+	sendAuthz, ok := authorization.(*banktypes.SendAuthorization)
+	if !ok {
+		return authz.ErrUnknownAuthorizationType
+	}
+
+	found, denomCoins := sendAuthz.SpendLimit.Find(p.tokenPair.Denom)
+	if !found {
+		return fmt.Errorf("allowance for token %s does not exist", p.tokenPair.Denom)
+	}
+
+	newSpendLimit, hasNeg := sendAuthz.SpendLimit.SafeSub(denomCoins)
+	if hasNeg {
+		return fmt.Errorf("subtracted value cannot be greater than existing allowance for denom %s: %s > %s",
```

Comment:
this should be covered too

Reply:
This can never occur since we are subtracting here what was found in the slice, so it's only a safety check here.